src/createRefArray/createRefArray.ts

```diff
@@ -0,0 +1,37 @@
+/**
+ * Trim values from end of array
+ */
+function trimEnd<T>(target: Array<T>, value: T): Array<T> {
+  for (let index = target.length - 1; index >= 0; index--) {
+    if (target[index] === value) {
+      target.pop();
+    } else {
+      break;
+    }
+  }
+
+  return target;
+}
+
+/**
+ * Creates an array that trims itself after it changes
+ */
+export function createRefArray<T extends Array<unknown>>(
+  initialTarget: T = [] as unknown as T,
+): typeof initialTarget {
+  let timeout = 0 as unknown as NodeJS.Timeout;
+
+  return new Proxy<typeof initialTarget>(trimEnd(initialTarget, null) as T, {
+    set(target, parameter, newValue): boolean {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-expect-error
+      target[parameter] = newValue;
+
+      // Use timeouts to minimize array mutations
+      clearTimeout(timeout);
+      timeout = setTimeout(() => trimEnd(target, null), 0);
```

Comment:
Added the `null` check. I've removed the timers and added this check, the timers are a bit overkill if we only trim it when indexes are set to `null`

Reply:
If a big list is removed it will still trim the array x amount of times. But the check is pretty quick until the last index is removed and the whole array is trimmed in 1 pass. So should be fine.